packages/pds/src/app-migrations/votes-to-likes.ts

```diff
@@ -0,0 +1,209 @@
+import { RepoService } from '../services/repo'
+
+const MIGRATION_NAME = '2023-03-15-votes-to-likes'
+const SHORT_NAME = 'votes-to-likes'
+const VOTE_LEX_ID = 'app.bsky.feed.vote'
+
+export async function votesToLikesMigration(ctx: AppContext) {
+  await appMigration(ctx.db, MIGRATION_NAME, (tx) => main(tx, ctx))
+}
+
+async function main(tx: Database, ctx: AppContext) {
+  console.log(SHORT_NAME, 'beginning')
+  tx.assertTransaction()
+  const { ref } = tx.db.dynamic
+  const now = new Date().toISOString()
+
+  // The message dispatcher usually ensures that the app view indexes these updates,
+  // but that has been taken care of with a db migration, plus the end of this migration.
+  const noopDispatcher = new MessageDispatcher()
+  noopDispatcher.destroy()
+
+  const repoTx = new RepoService(
+    tx,
+    ctx.repoSigningKey,
+    noopDispatcher,
+    ctx.blobstore,
+  )
+
+  // For each user remove vote records and replace with new like records.
+
+  const [didsWithVotes, voteCount] = await Promise.all([
+    getDidsWithVotes(tx),
+    getVoteCount(tx),
+  ])
+
+  console.log(
+    SHORT_NAME,
+    `${voteCount} votes across ${didsWithVotes.length} dids`,
+  )
+
+  let didsComplete = 0
+  let votesComplete = 0
+  const createsTurnedDeletes: string[] = []
+  const chunks = chunkArray(didsWithVotes, Math.ceil(didsWithVotes.length / 5))
+
+  await Promise.all(
+    chunks.map(async (dids) => {
+      for (const did of dids) {
+        const votes = await getVotes(tx, did)
+        console.log(SHORT_NAME, `${did} processing ${votes.length} votes`)
+
+        const writeResults = await Promise.all(
+          votes.map(async (vote) => {
+            const record = cborBytesToRecord(vote.bytes)
+            const del = prepareDelete({
+              did,
+              collection: VOTE_LEX_ID,
+              rkey: vote.rkey,
+            })
+
+            if (record.direction !== 'up') {
+              return del // Delete downvotes
+            }
+
+            try {
+              record['$type'] = ids.AppBskyFeedLike
+              delete record['direction']
+              assertValidRecord(record)
+            } catch {
+              // Delete invalid votes that can't be mapped to a like
+              createsTurnedDeletes.push(del.uri.toString())
+              return del
+            }
+
+            // Delete vote and replace it with a valid like record
+            return [
+              del,
+              await prepareCreate({
+                did,
+                collection: ids.AppBskyFeedLike,
+                rkey: vote.rkey, // Maintain same rkey
+                record,
+                validate: false, // Validated above
+              }),
+            ]
+          }),
+        )
+
+        // Chunk to avoid hitting e.g. postgres param limits
+        for (const writeChunk of chunkArray(writeResults, 1000)) {
+          await repoTx.processWrites(did, writeChunk.flat(), now)
```

Comment:
can't imagine it hitting this case, but maybe still worth ensuring that the chunk.length > 0